pytential/qbx/refinement.py

```diff
@@ -537,6 +537,28 @@ def _make_quad_stage2_discr(lpot_source, stage2_density_discr):
             lpot_source.real_dtype)
 
 
+def _make_temporary_collection(lpot_source,
+        stage1_density_discr=None,
+        stage2_density_discr=None):
+    from pytential import sym
+    from pytential import GeometryCollection
+
+    name = "_tmp_refine_source"
+    places = GeometryCollection(lpot_source, auto_where=name)
+
+    from pytential.symbolic.execution import _GEOMETRY_COLLECTION_DISCR_CACHE_NAME
+    discr_cache = places.get_cache(_GEOMETRY_COLLECTION_DISCR_CACHE_NAME)
+    if stage1_density_discr is not None:
+        discr_cache[(name, sym.QBX_SOURCE_STAGE1)] = stage1_density_discr
+
+    if stage2_density_discr is not None:
+        discr_cache[(name, sym.QBX_SOURCE_STAGE2)] = stage2_density_discr
+        discr_cache[(name, sym.QBX_SOURCE_QUAD_STAGE2)] = \
+                _make_quad_stage2_discr(lpot_source, stage2_density_discr)
```

Comment:
It would make me happier if this write access were done by a private method `_inject_discr_into_cache(geo_name, stage)` on the collection so that

- hash key generation can be somewhat central
- It's clear that users aren't supposed to poke around in (let alone write to) these caches

Reply:
Added a `_get_stuff_from_cache` and a `_add_stuff_to_cache` and used them everywhere else. Looks a lot nicer!